src/Functions/FunctionsConversion.cpp

```diff
@@ -4134,6 +4134,26 @@ class FunctionCast final : public IFunctionBase
         };
     }
 
+    /// Create wrapper only if we support this conversion.
+    WrapperType createWrapperIfCanConvert(const DataTypePtr & from, const DataTypePtr & to) const
+    {
+        try
+        {
+            /// We can avoid try/catch here if we will implement check that 2 types can be casted, but it
+            /// requires quite a lot of work. By now let's simply use try/catch.
+            /// First, check that we can create a wrapper.
+            WrapperType wrapper = prepareUnpackDictionaries(from, to);
+            /// Second, check if we can perform a conversion on empty columns.
+            ColumnsWithTypeAndName column_from = {{from->createColumn(), from, "" }};
+            wrapper(column_from, to, nullptr, 0);
+            return wrapper;
+        }
+        catch (...)
+        {
+            return {};
```

Comment:
Do we catch here only that exception?
`throw Exception(ErrorCodes::CANNOT_CONVERT_TYPE, "Cannot convert NULL to a non-nullable type");`

It would be more comfortable to read/understand if we narrow the scope of possible exceptions here.

Reply:
I beleive we can have these 3 error codes in this situation: `CANNOT_CONVERT_TYPE/ILLEGAL_TYPE_OF_ARGUMENT/ILLEGAL_COLUMN`.
For example: 
`Illegal type Array(Int64) of argument of function toUInt32`
` Illegal column Decimal32 of first argument of function toDateTime`
And some others

Reply:
Let me restrict it to these 3 cases and add more tests for them

Reply:
Ok, actually there are quite a lot of possible errors including also `NOT_IMPLEMENTED/CANNOT_PARSE_DATE/CANNOT_PARSE_DATETIME`. So let's just catch all `Exception` errors

Reply:
Also found one issue: https://github.com/ClickHouse/ClickHouse/issues/69078